lib/connection/msg.js

```diff
@@ -0,0 +1,293 @@
+    this.moreToCome = options.moreToCome || false;
+    this.exhaustAllowed = false;
+  }
+
+  toBin() {
+    const buffers = [];
+    let flags = 0;
+
+    if (this.checksumPresent) {
+      flags |= OPTS_CHECKSUM_PRESENT;
+    }
+
+    if (this.moreToCome) {
+      flags |= OPTS_MORE_TO_COME;
+    }
+
+    if (this.exhaustAllowed) {
+      flags |= OPTS_EXHAUST_ALLOWED;
+    }
+
+    const header = new Buffer(
+      4 * 4 + // Header
+        4 // Flags
+    );
+
+    buffers.push(header);
+
+    let totalLength = header.length;
+
+    const command = this.command;
+    const nameArgumentPair = validSegmentListNamePairs(command);
+    if (nameArgumentPair) {
+      // TODO: Add support for payload type 1
+      const argument = nameArgumentPair.argument;
+
+      // Add initial type 0 segment with arguments pulled up
+      const clonedQuery = Object.assign({}, command);
+      delete clonedQuery[argument];
+      totalLength += this.makeDocumentSegment(buffers, clonedQuery);
+
+      // Create type 1 query
+      totalLength += this.makeSequenceSegment(buffers, argument, command[argument]);
+    } else {
+      totalLength += this.makeDocumentSegment(buffers, command);
+    }
+
+    writeInt32ListToUint8Buffer(header, [totalLength, this.requestId, 0, opcodes.OP_MSG, flags]);
+
+    return buffers;
+  }
+
+  makeDocumentSegment(buffers, document) {
+    const payloadTypeBuffer = new Buffer(1);
+    payloadTypeBuffer[0] = 0;
+
+    const documentBuffer = this.serializeBson(document);
+
+    buffers.push(payloadTypeBuffer);
+    buffers.push(documentBuffer);
+
+    return payloadTypeBuffer.length + documentBuffer.length;
+  }
+
+  makeSequenceSegment(buffers, argument, documents) {
+    const metaBuffer = new Buffer(
+      1 + // payloadType,
+      4 + // Size of sequence
+      argument.length + // Argument length
+        1 //C string null terminator
+    );
+
+    let segmentLength = metaBuffer.length - 1;
+
+    buffers.push(metaBuffer);
+    documents.forEach(document => {
+      const documentBuffer = this.serializeBson(document);
+      segmentLength += documentBuffer.length;
+      buffers.push(documentBuffer);
+    });
+
+    metaBuffer[0] = 1 & 0x1;
+    metaBuffer[1] = segmentLength & 0xff;
+    metaBuffer[2] = (segmentLength >> 8) & 0xff;
+    metaBuffer[3] = (segmentLength >> 16) & 0xff;
+    metaBuffer[4] = (segmentLength >> 24) & 0xff;
+    metaBuffer.write(argument, 5, 'utf8');
+    metaBuffer[metaBuffer.length - 1] = 0;
+
+    return segmentLength + 1;
+  }
+
+  serializeBson(document) {
+    return this.bson.serialize(document, {
+      checkKeys: this.checkKeys,
+      serializeFunctions: this.serializeFunctions,
+      ignoreUndefined: this.ignoreUndefined
+    });
+  }
+}
+
+Msg.getRequestId = function() {
+  return ++_requestId;
+};
+
+class BinMsg {
+  constructor(bson, message, msgHeader, msgBody, opts) {
+    opts = opts || { promoteLongs: true, promoteValues: true, promoteBuffers: false };
```

Comment:
Can we document these options?

Reply:
Similarly, these are BSON-related options